### PR TITLE
ROX-20246: Update and cancel vuln exception v2 service

### PR DESCRIPTION
**Pull request by md2119**
## Description

Mistakenly deleted the remote branch of #8243.

This PR adds v2 service implementation for updating and canceling vuln exceptions. For deferral exceptions, only CVEs and expiry can be updated. For false-positive exceptions, only CVEs can be updated (because false positive exception do not have expiry). Updates to an enforced exception require approval. Until the update is approved, the original exception is enforced.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- ~Evaluated and added CHANGELOG entry if required~
- ~Determined and documented upgrade steps~
- ~Documented user facing changes~

If any of these don't apply, please comment below.

## Testing Performed
Unit test

### CODE REVIEW

File: proto/storage/vuln_requests.proto [hunk 1]
```diff
@@ -112,7 +121,9 @@ message VulnerabilityRequest {
 
     // 21 to 25 reserved for the updated request type oneof.
     oneof updated_req {
-        DeferralRequest              updated_deferral_req       = 21 [(gogoproto.moretags) = 'search:"-"'];
+        DeferralRequest              updated_deferral_req = 21 [deprecated = true, (gogoproto.moretags) = 'search:"-"'];
+        DeferralUpdate               deferral_update      = 22 [(gogoproto.moretags) = 'search:"-"'];
+        FalsePositiveUpdate          fp_update            = 23;
```
Comment (dashrews78) [inline proto/storage/vuln_requests.proto:126]: Is there a way to block 24-25 aside from this comment and hoping authors and reviewers see it?

Comment (dashrews78) [reply proto/storage/vuln_requests.proto:126]: nit (up to you):  for readability should we consider `false_positive_update`?

File: central/vulnerabilityrequest/datastore/datastore_impl.go [hunk 2]
```diff
@@ -231,6 +231,92 @@ func (ds *datastoreImpl) UpdateRequestExpiry(ctx context.Context, id, message st
...
+			if err != nil {
+				return nil, err
+			}
+			if req.Requestor.Id != user.UID() {
+				return nil, errors.Wrap(sac.ErrResourceAccessDenied, "requests can only be updated by the original requester")
+			}
+		}
+	}
+
+	if req.GetExpired() || req.GetStatus() == storage.RequestStatus_DENIED {
```
Comment (dashrews78) [inline central/vulnerabilityrequest/datastore/datastore_impl.go:267]: nit:  I would move this above the preceding SAC checks.  If the requests is expired or denied there is no point in doing that SAC check.

File: central/vulnerabilityrequest/manager/requestmgr/manager.go [hunk 3]
```diff
@@ -24,8 +24,13 @@ type Manager interface {
 	// Undo undoes and expires a vulnerability request. This action may not un-snooze the vulns in the scope impacted
 	// by the request if there are other active vul requests on the same cve affecting the same scope.
 	Undo(ctx context.Context, requestID string, requestParams *common.VulnRequestParams) (*storage.VulnerabilityRequest, error)
+	// Cancel cancels a vulnerability exception. Once canceled, the exception becomes and cannot be re-opened.
```
Comment (dashrews78) [inline central/vulnerabilityrequest/manager/requestmgr/manager.go:27]:
```suggestion
	// Cancel cancels a vulnerability exception. Once canceled, the exception cannot be re-opened.
```

File: central/vulnerabilityrequest/manager/requestmgr/manager_impl.go [hunk 4]
```diff
@@ -206,11 +208,45 @@ func (m *managerImpl) Undo(ctx context.Context, id string, _ *common.VulnRequest
 	return req, nil
 }
 
+func (m *managerImpl) Cancel(ctx context.Context, id string) (*storage.VulnerabilityRequest, error) {
+	if !features.UnifiedCVEDeferral.Enabled() {
+		return nil, errors.Errorf("set the environment variable %s=true and retry", features.UnifiedCVEDeferral.EnvVar())
+	}
+	if id == "" {
+		return nil, errox.InvalidArgs.CausedBy("ID must be specified")
```
Comment (dashrews78) [inline central/vulnerabilityrequest/manager/requestmgr/manager_impl.go:216]:
```suggestion
		return nil, errox.InvalidArgs.CausedBy("vulnerability exception ID must be specified")
```

File: central/vulnerabilityrequest/manager/requestmgr/manager_impl.go [hunk 5]
```diff
@@ -206,11 +208,45 @@ func (m *managerImpl) Undo(ctx context.Context, id string, _ *common.VulnRequest
...
+	return req, nil
+}
+
 func (m *managerImpl) UpdateExpiry(ctx context.Context, id string, reqParams *common.VulnRequestParams) (*storage.VulnerabilityRequest, error) {
-	// Currently, only deferral requests can be updated
+	if id == "" {
+		return nil, errox.InvalidArgs.CausedBy("vulnerability exception ID must be specified")
+	}
+	if reqParams == nil {
+		return nil, nil
```
Comment (dashrews78) [inline central/vulnerabilityrequest/manager/requestmgr/manager_impl.go:243]: This seems like an odd case.  Should it return an invalid args error too?

File: central/vulnerabilityrequest/service/v2/service_impl.go [hunk 6]
```diff
@@ -155,17 +155,48 @@ func (s *serviceImpl) DenyVulnerabilityException(ctx context.Context, req *v2.De
...
-// UpdateVulnerabilityException updates an existing vulnerability exception. Currently only the following can be updated:
+// UpdateVulnerabilityException updates an existing vulnerability exception. The update is enforced only once it is approved.
+// Currently, only the following can be updated:
 // - CVEs and expiration time of the deferral exceptions
 // - CVEs of the false positive requests
-func (s *serviceImpl) UpdateVulnerabilityException(_ context.Context, _ *v2.UpdateVulnerabilityExceptionRequest) (*v2.UpdateVulnerabilityExceptionResponse, error) {
-	return nil, errox.NotImplemented
+func (s *serviceImpl) UpdateVulnerabilityException(ctx context.Context, req *v2.UpdateVulnerabilityExceptionRequest) (*v2.UpdateVulnerabilityExceptionResponse, error) {
+	if req.GetId() == "" {
+		return nil, errox.InvalidArgs.CausedBy("ID not specified")
```
Comment (dashrews78) [inline central/vulnerabilityrequest/service/v2/service_impl.go:164]:
```suggestion
		return nil, errox.InvalidArgs.CausedBy("vulnerability exception ID must be specified")
```

File: central/vulnerabilityrequest/datastore/datastore_impl.go (hunk 2 above)
Comment (md2119) [reply central/vulnerabilityrequest/datastore/datastore_impl.go:267]: Actually no. We do not want to leak information about the request to the users if they do not have the required permission to complete the operation. Note that `req` is pulled out of the database to which the user may or may not have permission.

File: central/vulnerabilityrequest/manager/requestmgr/manager_impl.go (hunk 5 above)
Comment (md2119) [reply central/vulnerabilityrequest/manager/requestmgr/manager_impl.go:243]: good catch

File: proto/storage/vuln_requests.proto (hunk 1 above)
Comment (dashrews78) [reply proto/storage/vuln_requests.proto]: 🤞

**Review by dashrews78 — APPROVED** [review]

**Pull request merged**